### PR TITLE
Dedicated message grid

### PR DESCRIPTION
**Pull request by bfredl**
Problem: drawing messages on `default_grid` is messy with multigrid and floats
Solution: draw messages on dedicated grid.

Note: the current implementation is bullshit. We need to figure out the right events for screen.c to communicate with external UI, and internal compositor should use the same (and not use side channels).

But does it blend?
---

Try something like:

    hi MsgArea guibg=#102020 blend=20

### CODE REVIEW

**Comment by bfredl** [comment]
Note: it is also ridiculously slow with scrolling, especially with `blend=` . We should draw scrolling messages first internally, and only when `ui_flush` is called we will transmit the current state of `msg_grid` (which requires moving `ui_flush` from `msg_end` to relevant call places).

**Comment by bfredl** [comment]
Actually, removing `ui_flush()` from `msg_end()` doesn't seem to cause much trouble, probably because `ui_flush()` is always called before blocking the event loop. Perhaps we should add it back to some high-level message functions like `echo/echomsg/echoerr`, because a plugin could call them before going into busy CPU work, which would lead to no `ui_flush()`.

With that, throttling message drawing/scrolling doesn't seem too difficult. Compare

```
:set wd=10
:digraphs
```

on this branch with master (note: the time difference with `wd=10` is mostly bullshit, due to different event size. the relevant difference is the reduced scrolling).

Note: all of this assumes of course `display+=msgsep`. For `display-=msgsep` we should just alias `msg_grid` to `default_grid` and disable the throttling, and retain the vintage terminal scroll experience.

**Comment by bfredl** [comment]
Now this implements _one_ possible model for `ext_multigrid` with `ext_messages=false`. The message grid is an "overlay" grid which will cover the entire Rows X Columns area, but a `msg_set_pos(row)` event will indicate where the message area starts: the message grid should be only be drawn starting with this row, and above it the grid should not be drawn (and nvim will not send any grid_line events there). `msg_set_pos` will always be called before scrolling/drawing to make the message area bigger. This model is pretty much ideal for compositor/TUI: we use the same positioning logic and event kinds etc as before, just internal buffering to coalesce multiple scrolls into one big scroll.

The alternative (that maybe(?) suits external UIs better) would be to try make message grid look more like a float. I e `(0,0)` upper left corner of the message grid will correspond to the start of the message area, and instead of a `grid_scroll` there will be a `win_float_pos`-like event to move the relative position of the message grid. Then the part of the message grid that hangs outside the screen area should be considered invisible/invalid. (While strictly making messages.c using this model internally might require messy refactors, we could change the exported multigrid events to behave this way for now)

@akiyosi @vhakulinen and others what do you think? (I know gnvim is focusing on ext_messages now, but your insight may still be valuable)

**Comment by akiyosi** [comment]
@bfredl 
Thank you for the repeated support by gitter for this PR.
I apologize because I bothered you about my lack of understanding😓

In the current implementation, message grid settled on the implementation that treats like float window.
i.e. 
```
["msg_set_pos", msg_grid, start_row]
```
I confirmed that the behavior of the message grid works well with the external GUI.

In this implementation, I think that the cursor position on the command line should be reported as a position on message grid instead of the global grid, is my recognition correct?
It seems that the cursor position on the command line is still notified on the global grid now. Therefore, the cursor position may shift slightly.

**Comment by bfredl** [comment]
@akiyosi 

> I apologize because I bothered you about my lack of understanding

There is no need to apologize for anything, understanding is always made by communication.

>  In this implementation, I think that the cursor position on the command line should be reported as a position on message grid instead of the global grid, is my recognition correct?

This is correct, I will fix it.

**Comment by bfredl** [comment]
Note to self: check if this fixes #10262

**Comment by bfredl** [comment]
I think this is starting to get done. I have added documentation for the changes of the multigrid protocol.

The other user-facing change (apart from optimization) is the addition for a highlight for the background of the message/cmdline area. Currently I call this "Message", but maybe "MessageArea" is better?

File: src/nvim/api/ui_events.in.h [hunk 1]
```diff
@@ -112,9 +112,7 @@ void win_hide(Integer grid)
   FUNC_API_SINCE(6) FUNC_API_REMOTE_ONLY;
 void win_close(Integer grid)
   FUNC_API_SINCE(6) FUNC_API_REMOTE_ONLY;
-void win_scroll_over_start(void)
-  FUNC_API_SINCE(6) FUNC_API_BRIDGE_IMPL FUNC_API_COMPOSITOR_IMPL;
-void win_scroll_over_reset(void)
+void msg_set_pos(Integer grid, Integer row, Boolean scrolled, String sep_char)
   FUNC_API_SINCE(6) FUNC_API_BRIDGE_IMPL FUNC_API_COMPOSITOR_IMPL;
```
Comment (justinmk) [inline src/nvim/api/ui_events.in.h]: should it be grouped with the other msg_xx events (line 162)?

File: src/nvim/highlight_defs.h [hunk 2]
```diff
@@ -146,6 +147,7 @@ EXTERN const char *hlf_names[] INIT(= {
   [HLF_INACTIVE] = "NormalNC",
   [HLF_MSGSEP] = "MsgSeparator",
   [HLF_NFLOAT] = "NormalFloat",
+  [HLF_MSG] = "Message",
```
Comment (justinmk) [inline src/nvim/highlight_defs.h]: how about `MsgArea`, for parallel with `MsgSeparator`/`ErrorMsg`/etc.

File: src/nvim/screen.c [hunk 3]
```diff
@@ -5886,6 +5915,18 @@ void grid_fill(ScreenGrid *grid, int start_row, int end_row, int start_col,
       if (put_dirty_row == row) {
         put_dirty_first = MIN(put_dirty_first, dirty_first);
         put_dirty_last = MAX(put_dirty_last, dirty_last);
+      } else if (grid->throttled) {
+        // Note: assumes msg_grid is the only thottled grid
```
Comment (justinmk) [inline src/nvim/screen.c]:
```suggestion
        // Note: assumes msg_grid is the only throttled grid
```

File: src/nvim/api/ui_events.in.h (hunk 1 above)
Comment (bfredl) [reply src/nvim/api/ui_events.in.h]: No, the `msg_show` etc events are for `ext_messages`. These events here are the ones you need if you use `ext_multigrid` (and nothing else).

File: src/nvim/highlight_defs.h (hunk 2 above)
Comment (bfredl) [reply src/nvim/highlight_defs.h]: `MsgArea` sounds good :+1:

**Pull request merged**